_tools/types/main.go

```diff
@@ -0,0 +1,340 @@
+func maybeCloneOrPullThrottled(d *driverStats, concurent chan int) error {
+	concurent <- 1
+	defer func() {
+		<-concurent
+	}()
+
+	return maybeCloneOrPull(d)
+}
+
+func maybeCloneOrPull(d *driverStats) error {
+	repoPath := path.Join(reposRootPath, d.path)
+	_, err := os.Stat(repoPath)
+	if os.IsNotExist(err) {
+		fmt.Fprintf(os.Stderr, "%s does not exist, cloning from %s\n", repoPath, d.url)
+		cmd := exec.Command("git", "clone", d.url+".git")
+		cmd.Dir = reposRootPath
+		err = cmd.Run()
+		if err != nil {
+			return err
+		}
+		return nil
+	} else if err != nil {
+		fmt.Fprintf(os.Stderr, "error: %s\n", err)
+		return err
+	}
+
+	fmt.Fprintf(os.Stderr, "%s dir exists, will git pull instead\n", repoPath)
+	cmd := exec.Command("git", "pull", "origin", "master")
+	cmd.Dir = repoPath
+	err = cmd.Run()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type uastType struct {
+	name string
+}
+
+// isUsedIn checks if current UAST type is used in the given package.
+func (u *uastType) isUsedIn() {
+	// TODO
+}
+
+// findUastTypesInSdk finds all types from SDK.
+func findUastTypesInSdk() []uastType {
+	var out []uastType
+	// TODO: load package, iterate all structs
+	types := []interface{}{
+		uast.Position{},
+		uast.Positions{},
+		// only embeding GenNode?
+		uast.Identifier{},
+		uast.String{},
+		uast.Bool{},
+		uast.QualifiedIdentifier{},
+		uast.Comment{},
+		uast.Group{},
+		uast.FunctionGroup{},
+		uast.Block{},
+		uast.Alias{},
+		uast.Import{},
+		uast.RuntimeImport{},
+		uast.RuntimeReImport{},
+		uast.InlineImport{},
+		uast.Argument{},
+		uast.FunctionType{},
+		uast.Function{},
+	}
+	for _, typee := range types {
+		out = append(out, uastType{reflect.TypeOf(typee).String()})
+	}
+	fmt.Fprintf(os.Stderr, "%d uast:* types found in SDK\n", len(out))
+	return out
+}
+
+// analyzeFixtures goes though all fixtures, assuming the driver is cloned.
+// It updates given driverStats with results.
+func analyzeFixtures(driver *driverStats) error {
+	fixDir := path.Join(reposRootPath, driver.path, fixtureDir)
+	fixtureFiles, err := lsDir(fixDir)
+	if os.IsNotExist(err) {
+		driver.skip = true
+		return nil
+	} else if err != nil {
+		return err
+	}
+	driver.fixturesNum += len(fixtureFiles)
+
+	for _, file := range fixtureFiles {
+		if !strings.HasSuffix(file.Name(), fixtureExt) {
+			continue
+		}
+		
+		fFile, err := os.Open(path.Join(fixDir, file.Name()))
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer fFile.Close()
+
+		scanner := bufio.NewScanner(fFile)
+		for scanner.Scan() {
+			line := scanner.Text()
+			if !strings.Contains(line, "uast:") {
+				continue
+			}
+			for _, uastType := range uastNodeRe.FindAllString(line, -1) {
+				driver.fixturesUast[strings.Replace(uastType, ":", ".", 1)]++
+			}
+		}
+
+		if err := scanner.Err(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func lsDir(dir string) ([]os.FileInfo, error) {
+	fmt.Fprintf(os.Stderr, "reading %s/*%s files\n", dir, fixtureExt)
+	f, err := os.Open(dir)
```

Comment:
https://godoc.org/io/ioutil#ReadDir may help here.

Reply:
I'm aware about [ReadDir](https://golang.org/src/io/ioutil/ioutil.go#L96) but it sorts the output. In our case, we do not care about an order and thus avoid wasting some CPU cycles.

Reply:
Okay, but I don't think these directories are very big, and this doesn't seem like a performance critical code path. If it does matter, though, I'd be fine with adding a comment to mention why we are bypassing the simpler solution. 😄

Reply:
added comment to go doc